<a href="https://colab.research.google.com/github/gauriagarwal18/NYC-Taxi-Trip-Time-Prediction/blob/master/NYC_Taxi_Trip_Time_Prediction_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <b><u> Project Title : Taxi trip time Prediction : Predicting total ride duration of taxi trips in New York City</u></b>

## <b> Problem Description </b>

### Your task is to build a model that predicts the total ride duration of taxi trips in New York City. Your primary dataset is one released by the NYC Taxi and Limousine Commission, which includes pickup time, geo-coordinates, number of passengers, and several other variables.

##We are using the following libraries for analysis:
- Numpy: We will use numpy arrays as they are comparitively faster than lists, also columns of dataframes behaves as numpy arrays

- Pandas: for reading the data from csv file, for data clening and for preparing data for analysis

- matplotlib,seaborn: for different visualisations, for drawing conclusions from data and for exploratory data analysis. 

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

##Loading CSV file from google drive in colab enviroment
We will first load our csv file in which data is stored to the colab enviroment in data frame format so that we can make the copy of the original data and perform the required cleaning and analysis on that data without changing the original one.



In [9]:
#Download datasets Hepatitis automobile from UCI repository
from google.colab import drive
drive.mount('/content/drive')
import os
path="/content/drive/My Drive/AlmaBetter_Capstone_projects/Capstone_project2_ml/NYC_TaxiData.csv"

taxi_original=pd.read_csv(path,parse_dates=[2,3])
taxi= taxi_original.copy()

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [10]:
#shape of the data
taxi.shape

(1458644, 11)

In [11]:
taxi.head()

id  vendor_id     pickup_datetime    dropoff_datetime  \
0  id2875421          2 2016-03-14 17:24:55 2016-03-14 17:32:30   
1  id2377394          1 2016-06-12 00:43:35 2016-06-12 00:54:38   
2  id3858529          2 2016-01-19 11:35:24 2016-01-19 12:10:48   
3  id3504673          2 2016-04-06 19:32:31 2016-04-06 19:39:40   
4  id2181028          2 2016-03-26 13:30:55 2016-03-26 13:38:10   

   passenger_count  pickup_longitude  pickup_latitude  dropoff_longitude  \
0                1        -73.982155        40.767937         -73.964630   
1                1        -73.980415        40.738564         -73.999481   
2                1        -73.979027        40.763939         -74.005333   
3                1        -74.010040        40.719971         -74.012268   
4                1        -73.973053        40.793209         -73.972923   

   dropoff_latitude store_and_fwd_flag  trip_duration  
0         40.765602                  N            455  
1         40.731152                  N            663  
2         40.710087                  N           2124  
3         40.706718                  N            429  
4         40.782520                  N            435

## <b> Data Description </b>

### The dataset is based on the 2016 NYC Yellow Cab trip record data made available in Big Query on Google Cloud Platform. The data was originally published by the NYC Taxi and Limousine Commission (TLC). The data was sampled and cleaned for the purposes of this project. Based on individual trip attributes, you should predict the duration of each trip in the test set.

### <b>NYC Taxi Data.csv</b> - the training set (contains 1458644 trip records)


### Data fields
* #### id - a unique identifier for each trip
* #### vendor_id - a code indicating the provider associated with the trip record
* #### pickup_datetime - date and time when the meter was engaged
* #### dropoff_datetime - date and time when the meter was disengaged
* #### passenger_count - the number of passengers in the vehicle (driver entered value)
* #### pickup_longitude - the longitude where the meter was engaged
* #### pickup_latitude - the latitude where the meter was engaged
* #### dropoff_longitude - the longitude where the meter was disengaged
* #### dropoff_latitude - the latitude where the meter was disengaged
* #### store_and_fwd_flag - This flag indicates whether the trip record was held in vehicle memory before sending to the vendor because the vehicle did not have a connection to the server - Y=store and forward; N=not a store and forward trip
* #### trip_duration - duration of the trip in seconds

In [12]:
taxi.info()
"""
here we note that there is not any null value in data right now
"""

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1458644 entries, 0 to 1458643
Data columns (total 11 columns):
 #   Column              Non-Null Count    Dtype         
---  ------              --------------    -----         
 0   id                  1458644 non-null  object        
 1   vendor_id           1458644 non-null  int64         
 2   pickup_datetime     1458644 non-null  datetime64[ns]
 3   dropoff_datetime    1458644 non-null  datetime64[ns]
 4   passenger_count     1458644 non-null  int64         
 5   pickup_longitude    1458644 non-null  float64       
 6   pickup_latitude     1458644 non-null  float64       
 7   dropoff_longitude   1458644 non-null  float64       
 8   dropoff_latitude    1458644 non-null  float64       
 9   store_and_fwd_flag  1458644 non-null  object        
 10  trip_duration       1458644 non-null  int64         
dtypes: datetime64[ns](2), float64(4), int64(3), object(2)
memory usage: 122.4+ MB


'\nhere we note that there is not any null value in data right now\n'

In [7]:
taxi.describe(include="all")

id     vendor_id      pickup_datetime     dropoff_datetime  \
count     1458644  1.458644e+06              1458644              1458644   
unique    1458644           NaN              1380222              1380377   
top     id2875421           NaN  2016-03-04 08:07:34  2016-02-19 19:25:04   
freq            1           NaN                    5                    5   
mean          NaN  1.534950e+00                  NaN                  NaN   
std           NaN  4.987772e-01                  NaN                  NaN   
min           NaN  1.000000e+00                  NaN                  NaN   
25%           NaN  1.000000e+00                  NaN                  NaN   
50%           NaN  2.000000e+00                  NaN                  NaN   
75%           NaN  2.000000e+00                  NaN                  NaN   
max           NaN  2.000000e+00                  NaN                  NaN   

        passenger_count  pickup_longitude  pickup_latitude  dropoff_longitude  \
count      1.458644e+06      1.458644e+06     1.458644e+06       1.458644e+06   
unique              NaN               NaN              NaN                NaN   
top                 NaN               NaN              NaN                NaN   
freq                NaN               NaN              NaN                NaN   
mean       1.664530e+00     -7.397349e+01     4.075092e+01      -7.397342e+01   
std        1.314242e+00      7.090186e-02     3.288119e-02       7.064327e-02   
min        0.000000e+00     -1.219333e+02     3.435970e+01      -1.219333e+02   
25%        1.000000e+00     -7.399187e+01     4.073735e+01      -7.399133e+01   
50%        1.000000e+00     -7.398174e+01     4.075410e+01      -7.397975e+01   
75%        2.000000e+00     -7.396733e+01     4.076836e+01      -7.396301e+01   
max        9.000000e+00     -6.133553e+01     5.188108e+01      -6.133553e+01   

        dropoff_latitude store_and_fwd_flag  trip_duration  
count       1.458644e+06            1458644   1.458644e+06  
unique               NaN                  2            NaN  
top                  NaN                  N            NaN  
freq                 NaN            1450599            NaN  
mean        4.075180e+01                NaN   9.594923e+02  
std         3.589056e-02                NaN   5.237432e+03  
min         3.218114e+01                NaN   1.000000e+00  
25%         4.073588e+01                NaN   3.970000e+02  
50%         4.075452e+01                NaN   6.620000e+02  
75%         4.076981e+01                NaN   1.075000e+03  
max         4.392103e+01                NaN   3.526282e+06

In [14]:
taxi.columns

Index(['id', 'vendor_id', 'pickup_datetime', 'dropoff_datetime',
       'passenger_count', 'pickup_longitude', 'pickup_latitude',
       'dropoff_longitude', 'dropoff_latitude', 'store_and_fwd_flag',
       'trip_duration'],
      dtype='object')

In [16]:
categorical=["vendor_id","passenger_count","store_and_fwd_flag"]
for c in categorical:
  print(f"distribution of {c}:\n{taxi[c].value_counts()}\n\n")

distribution of vendor_id:
2    780302
1    678342
Name: vendor_id, dtype: int64


distribution of passenger_count:
1    1033540
2     210318
5      78088
3      59896
6      48333
4      28404
0         60
7          3
9          1
8          1
Name: passenger_count, dtype: int64


distribution of store_and_fwd_flag:
N    1450599
Y       8045
Name: store_and_fwd_flag, dtype: int64




In [22]:
(taxi[taxi["passenger_count"]==0]).loc["passenger_count"]=np.nan

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:723: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value, self.name)


In [23]:
taxi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1458644 entries, 0 to 1458643
Data columns (total 11 columns):
 #   Column              Non-Null Count    Dtype         
---  ------              --------------    -----         
 0   id                  1458644 non-null  object        
 1   vendor_id           1458644 non-null  int64         
 2   pickup_datetime     1458644 non-null  datetime64[ns]
 3   dropoff_datetime    1458644 non-null  datetime64[ns]
 4   passenger_count     1458644 non-null  int64         
 5   pickup_longitude    1458644 non-null  float64       
 6   pickup_latitude     1458644 non-null  float64       
 7   dropoff_longitude   1458644 non-null  float64       
 8   dropoff_latitude    1458644 non-null  float64       
 9   store_and_fwd_flag  1458644 non-null  object        
 10  trip_duration       1458644 non-null  int64         
dtypes: datetime64[ns](2), float64(4), int64(3), object(2)
memory usage: 122.4+ MB


In [ ]:
#the data with passenger count=0 is of no use so we will delete those rows:
taxi.drop()